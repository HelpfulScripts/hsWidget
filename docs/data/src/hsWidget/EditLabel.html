
                <html><style>body { overflow:hidden;}
            h1 { font-family: Arial, sans-serif; font-size: 24px; color: #44a; }
            p { margin:0; padding:0; }
            br  { margin:0; padding:0; }
            .line { 
                white-space: nowrap;
                height:16px; 
            }
            .line>span { 
                display:inline-block; background-color:#eef; height:100%; 
                margin: 0 5px 0 0; padding-right: 5px; color:#999;   
            }
            comment {color: #080;} module {color: #804;} 
            quote {color: #008;} comment>quote {color: #080;}
            .listing { margin: 10px; border: 1px solid #ccc; 
                    font-family: SFMono-Regular, Consolas, 'Liberation Mono', Menlo, Courier, monospace;
                    font-size: 14px;  
                    overflow:scroll;
                    height:90%;
            }
            code { padding: 5px 0;}</style></html><body>
                <h1>EditLabel.ts</h1>
                <div class='listing'><code><p id=1 class="line"><span>&nbsp;&nbsp;&nbsp;1</span><comment>/**</comment></p>
<p id=2 class="line"><span>&nbsp;&nbsp;&nbsp;2</span><comment>&nbsp;*&nbsp;#&nbsp;EditLabel</comment></p>
<p id=3 class="line"><span>&nbsp;&nbsp;&nbsp;3</span><comment>&nbsp;*&nbsp;Provides&nbsp;a&nbsp;text&nbsp;label&nbsp;that&nbsp;turns&nbsp;into&nbsp;a&nbsp;single-line&nbsp;input&nbsp;field&nbsp;when&nbsp;clicked&nbsp;on</comment></p>
<p id=4 class="line"><span>&nbsp;&nbsp;&nbsp;4</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=5 class="line"><span>&nbsp;&nbsp;&nbsp;5</span><comment>&nbsp;*&nbsp;###&nbsp;Attributes:</comment></p>
<p id=6 class="line"><span>&nbsp;&nbsp;&nbsp;6</span><comment>&nbsp;*&nbsp;-&nbsp;**css**?:&nbsp;`string`&nbsp;an&nbsp;optional&nbsp;css&nbsp;selector</comment></p>
<p id=7 class="line"><span>&nbsp;&nbsp;&nbsp;7</span><comment>&nbsp;*&nbsp;-&nbsp;**content**:&nbsp;`string`&nbsp;the&nbsp;initial&nbsp;string&nbsp;to&nbsp;show&nbsp;in&nbsp;the&nbsp;text&nbsp;label</comment></p>
<p id=8 class="line"><span>&nbsp;&nbsp;&nbsp;8</span><comment>&nbsp;*&nbsp;-&nbsp;**update**:&nbsp;`(newValue:string)=>void`&nbsp;callback&nbsp;function&nbsp;called&nbsp;with&nbsp;the&nbsp;updated&nbsp;lable&nbsp;when&nbsp;the&nbsp;input&nbsp;field&nbsp;loses&nbsp;focus</comment></p>
<p id=9 class="line"><span>&nbsp;&nbsp;&nbsp;9</span><comment>&nbsp;*&nbsp;-&nbsp;**placeholder**?:&nbsp;optional&nbsp;hint&nbsp;to&nbsp;show&nbsp;for&nbsp;an&nbsp;empty&nbsp;field.&nbsp;Deafults&nbsp;to&nbsp;*<quote>'click&nbsp;to&nbsp;enter'</quote>*.&nbsp;Provide&nbsp;a&nbsp;`sapce&nbsp;(<quote>'&nbsp;'</quote>)`&nbsp;to&nbsp;</comment></p>
<p id=10 class="line"><span>&nbsp;&nbsp;10</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;display&nbsp;an&nbsp;empty&nbsp;field.</comment></p>
<p id=11 class="line"><span>&nbsp;&nbsp;11</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=12 class="line"><span>&nbsp;&nbsp;12</span><comment>&nbsp;*&nbsp;###&nbsp;Example</comment></p>
<p id=13 class="line"><span>&nbsp;&nbsp;13</span><comment>&nbsp;*&nbsp;<example></comment></p>
<p id=14 class="line"><span>&nbsp;&nbsp;14</span><comment>&nbsp;*&nbsp;<file&nbsp;name=<quote>'script.js'</quote>>&nbsp;</comment></p>
<p id=15 class="line"><span>&nbsp;&nbsp;15</span><comment>&nbsp;*&nbsp;let&nbsp;content&nbsp;=&nbsp;<quote>''</quote>;</comment></p>
<p id=16 class="line"><span>&nbsp;&nbsp;16</span><comment>&nbsp;*&nbsp;m.mount(root,&nbsp;{view:&nbsp;()&nbsp;=>&nbsp;m(<quote>'div'</quote>,&nbsp;[</comment></p>
<p id=17 class="line"><span>&nbsp;&nbsp;17</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'h3'</quote>,&nbsp;<quote>'Edit&nbsp;Label&nbsp;Example'</quote>),</comment></p>
<p id=18 class="line"><span>&nbsp;&nbsp;18</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'div'</quote>,&nbsp;[</comment></p>
<p id=19 class="line"><span>&nbsp;&nbsp;19</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(<quote>'span'</quote>,&nbsp;`current&nbsp;content:&nbsp;<quote>'${content}'</quote>`),</comment></p>
<p id=20 class="line"><span>&nbsp;&nbsp;20</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(hsWidget.EditLabel,&nbsp;{</comment></p>
<p id=21 class="line"><span>&nbsp;&nbsp;21</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;css:&nbsp;<quote>'.myLabel'</quote>,</comment></p>
<p id=22 class="line"><span>&nbsp;&nbsp;22</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;placeholder:&nbsp;<quote>'Enter&nbsp;Value'</quote>,</comment></p>
<p id=23 class="line"><span>&nbsp;&nbsp;23</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;content:&nbsp;content,</comment></p>
<p id=24 class="line"><span>&nbsp;&nbsp;24</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;update:&nbsp;newValue&nbsp;=>&nbsp;content&nbsp;=&nbsp;newValue</comment></p>
<p id=25 class="line"><span>&nbsp;&nbsp;25</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;})</comment></p>
<p id=26 class="line"><span>&nbsp;&nbsp;26</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;])</comment></p>
<p id=27 class="line"><span>&nbsp;&nbsp;27</span><comment>&nbsp;*&nbsp;])});</comment></p>
<p id=28 class="line"><span>&nbsp;&nbsp;28</span><comment>&nbsp;*&nbsp;</comment></p>
<p id=29 class="line"><span>&nbsp;&nbsp;29</span><comment>&nbsp;*&nbsp;</file></comment></p>
<p id=30 class="line"><span>&nbsp;&nbsp;30</span><comment>&nbsp;*&nbsp;<file&nbsp;name=<quote>'style.css'</quote>></comment></p>
<p id=31 class="line"><span>&nbsp;&nbsp;31</span><comment>&nbsp;*&nbsp;.myLabel&nbsp;{&nbsp;&nbsp;</comment></p>
<p id=32 class="line"><span>&nbsp;&nbsp;32</span><comment>&nbsp;*&nbsp;}</comment></p>
<p id=33 class="line"><span>&nbsp;&nbsp;33</span><comment>&nbsp;*&nbsp;</file></comment></p>
<p id=34 class="line"><span>&nbsp;&nbsp;34</span><comment>&nbsp;*&nbsp;</example></comment></p>
<p id=35 class="line"><span>&nbsp;&nbsp;35</span><comment>&nbsp;*/</comment></p>
<p id=36 class="line"><span>&nbsp;&nbsp;36</span></p>
<p id=37 class="line"><span>&nbsp;&nbsp;37</span><comment>/**&nbsp;*/</comment></p>
<p id=38 class="line"><span>&nbsp;&nbsp;38</span>import&nbsp;{&nbsp;Log&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;<quote>'hsutil'</quote>;&nbsp;&nbsp;const&nbsp;log&nbsp;=&nbsp;new&nbsp;Log(<quote>'EditLabel'</quote>);</p>
<p id=39 class="line"><span>&nbsp;&nbsp;39</span>import&nbsp;{&nbsp;m,&nbsp;Vnode&nbsp;}&nbsp;from&nbsp;<quote>'hslayout'</quote>;</p>
<p id=40 class="line"><span>&nbsp;&nbsp;40</span></p>
<p id=41 class="line"><span>&nbsp;&nbsp;41</span></p>
<p id=42 class="line"><span>&nbsp;&nbsp;42</span>export&nbsp;class&nbsp;EditLabel&nbsp;{</p>
<p id=43 class="line"><span>&nbsp;&nbsp;43</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;editable&nbsp;=&nbsp;false;</p>
<p id=44 class="line"><span>&nbsp;&nbsp;44</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;hasFocus&nbsp;=&nbsp;false;</p>
<p id=45 class="line"><span>&nbsp;&nbsp;45</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;updateCB:(r:string)&nbsp;=>&nbsp;void;</p>
<p id=46 class="line"><span>&nbsp;&nbsp;46</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;default&nbsp;=&nbsp;<quote>''</quote>;</p>
<p id=47 class="line"><span>&nbsp;&nbsp;47</span></p>
<p id=48 class="line"><span>&nbsp;&nbsp;48</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;click(e:any)&nbsp;{&nbsp;</p>
<p id=49 class="line"><span>&nbsp;&nbsp;49</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.editable&nbsp;=&nbsp;!this.editable;&nbsp;</p>
<p id=50 class="line"><span>&nbsp;&nbsp;50</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=51 class="line"><span>&nbsp;&nbsp;51</span></p>
<p id=52 class="line"><span>&nbsp;&nbsp;52</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;blur(e:any)&nbsp;{</p>
<p id=53 class="line"><span>&nbsp;&nbsp;53</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.editable&nbsp;=&nbsp;false;</p>
<p id=54 class="line"><span>&nbsp;&nbsp;54</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.hasFocus&nbsp;=&nbsp;false;</p>
<p id=55 class="line"><span>&nbsp;&nbsp;55</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.update(e.target.value);</p>
<p id=56 class="line"><span>&nbsp;&nbsp;56</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=57 class="line"><span>&nbsp;&nbsp;57</span></p>
<p id=58 class="line"><span>&nbsp;&nbsp;58</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;keyup(key:any)&nbsp;{</p>
<p id=59 class="line"><span>&nbsp;&nbsp;59</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if&nbsp;(key.which&nbsp;===&nbsp;13)&nbsp;{</p>
<p id=60 class="line"><span>&nbsp;&nbsp;60</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;key.target.blur();</p>
<p id=61 class="line"><span>&nbsp;&nbsp;61</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=62 class="line"><span>&nbsp;&nbsp;62</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=63 class="line"><span>&nbsp;&nbsp;63</span></p>
<p id=64 class="line"><span>&nbsp;&nbsp;64</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;update(newValue:string)&nbsp;{</p>
<p id=65 class="line"><span>&nbsp;&nbsp;65</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.updateCB(newValue);</p>
<p id=66 class="line"><span>&nbsp;&nbsp;66</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=67 class="line"><span>&nbsp;&nbsp;67</span></p>
<p id=68 class="line"><span>&nbsp;&nbsp;68</span>&nbsp;&nbsp;&nbsp;&nbsp;public&nbsp;onupdate(node:Vnode)&nbsp;{</p>
<p id=69 class="line"><span>&nbsp;&nbsp;69</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if&nbsp;(this.editable&nbsp;&&&nbsp;!this.hasFocus)&nbsp;{</p>
<p id=70 class="line"><span>&nbsp;&nbsp;70</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;node.dom.value&nbsp;=&nbsp;node.attrs.content&nbsp;||&nbsp;this.default;</p>
<p id=71 class="line"><span>&nbsp;&nbsp;71</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;node.dom.focus();</p>
<p id=72 class="line"><span>&nbsp;&nbsp;72</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;node.dom.select();</p>
<p id=73 class="line"><span>&nbsp;&nbsp;73</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.hasFocus&nbsp;=&nbsp;true;</p>
<p id=74 class="line"><span>&nbsp;&nbsp;74</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=75 class="line"><span>&nbsp;&nbsp;75</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=76 class="line"><span>&nbsp;&nbsp;76</span></p>
<p id=77 class="line"><span>&nbsp;&nbsp;77</span>&nbsp;&nbsp;&nbsp;&nbsp;public&nbsp;view(node:Vnode)&nbsp;{</p>
<p id=78 class="line"><span>&nbsp;&nbsp;78</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.updateCB&nbsp;=&nbsp;node.attrs.update;</p>
<p id=79 class="line"><span>&nbsp;&nbsp;79</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;css&nbsp;=&nbsp;node.attrs.css&nbsp;||&nbsp;<quote>''</quote>;</p>
<p id=80 class="line"><span>&nbsp;&nbsp;80</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;this.editable?&nbsp;</p>
<p id=81 class="line"><span>&nbsp;&nbsp;81</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(`input.hsedit_label${css}`,&nbsp;{&nbsp;</p>
<p id=82 class="line"><span>&nbsp;&nbsp;82</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;onblur:&nbsp;this.blur.bind(this),</p>
<p id=83 class="line"><span>&nbsp;&nbsp;83</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;onkeyup:&nbsp;this.keyup.bind(this),</p>
<p id=84 class="line"><span>&nbsp;&nbsp;84</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;},<quote>''</quote>)</p>
<p id=85 class="line"><span>&nbsp;&nbsp;85</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;:&nbsp;(node.attrs.content&nbsp;&&&nbsp;node.attrs.content.length)?&nbsp;</p>
<p id=86 class="line"><span>&nbsp;&nbsp;86</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;m(`span.hsedit_label${css}`,&nbsp;{&nbsp;onclick:&nbsp;this.click.bind(this)&nbsp;},&nbsp;m.trust(node.attrs.content))</p>
<p id=87 class="line"><span>&nbsp;&nbsp;87</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;:&nbsp;m(`span.hsedit_label.default${css}`,&nbsp;{&nbsp;onclick:&nbsp;this.click.bind(this)&nbsp;},&nbsp;node.attrs.placeholder&nbsp;||&nbsp;<quote>'click&nbsp;to&nbsp;enter'</quote>);</p>
<p id=88 class="line"><span>&nbsp;&nbsp;88</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=89 class="line"><span>&nbsp;&nbsp;89</span>}</p>
<p id=90 class="line"><span>&nbsp;&nbsp;90</span></p>
<p id=91 class="line"><span>&nbsp;&nbsp;91</span>export&nbsp;class&nbsp;EditDate&nbsp;extends&nbsp;EditLabel&nbsp;{</p>
<p id=92 class="line"><span>&nbsp;&nbsp;92</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;default&nbsp;=&nbsp;new&nbsp;Date().toDateString().slice(4);</p>
<p id=93 class="line"><span>&nbsp;&nbsp;93</span>&nbsp;&nbsp;&nbsp;&nbsp;protected&nbsp;update(newValue:string)&nbsp;{</p>
<p id=94 class="line"><span>&nbsp;&nbsp;94</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if&nbsp;(newValue)&nbsp;{</p>
<p id=95 class="line"><span>&nbsp;&nbsp;95</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;date&nbsp;=&nbsp;new&nbsp;Date(newValue);</p>
<p id=96 class="line"><span>&nbsp;&nbsp;96</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;result&nbsp;=&nbsp;isNaN(date.getTime())?&nbsp;<quote>'invalid&nbsp;date'</quote>&nbsp;:&nbsp;date.toDateString().slice(4);</p>
<p id=97 class="line"><span>&nbsp;&nbsp;97</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;super.update(result);</p>
<p id=98 class="line"><span>&nbsp;&nbsp;98</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}&nbsp;else&nbsp;{</p>
<p id=99 class="line"><span>&nbsp;&nbsp;99</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;super.update(undefined);</p>
<p id=100 class="line"><span>&nbsp;100</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=101 class="line"><span>&nbsp;101</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=102 class="line"><span>&nbsp;102</span>}</p></code></div>
                </body>
            