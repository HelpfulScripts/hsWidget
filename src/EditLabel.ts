/**
 * # EditLabel
 * Provides a text label that turns into a single-line input field when clicked on
 * 
 * ### Attributes:
 * - **css**?: `string` an optional css selector
 * - **content**: `string` the initial string to show in the text label
 * - **update**: `(newValue:string)=>void` callback function called with the updated lable when the input field loses focus
 * - **popup**?: `string`: optional message to show in a {@link Popup popup box} upon mouseover. 
 * - **placeholder**?: optional hint to show for an empty field. Deafults to *'click to enter'*. Provide a `sapce (' ')` to 
 *      display an empty field.
 * 
 * ### Example
 * <example>
 * <file name='script.js'> 
 * let content = 'prior content';
 * m.mount(root, {view: () => m('div', [
 *    m('h3', 'Edit Label Example'),
 *    m('div', [
 *      m('span', `current content: '${content}'`),
 *      m(hsWidget.EditLabel, {
 *          css: '.myLabel',
 *          placeholder: 'Enter Value',
 *          content: content,
 *          update: newValue => content = newValue
 *      })
 *    ])
 * ])});
 * 
 * </file>
 * <file name='style.css'>
 * span.myLabel { background-color: #eee; }
 * input.myLabel { background-color: #fff; }
 * </file>
 * </example>
 */

/** */
import { Log }      from 'hsutil';  const log = new Log('EditLabel');
import { m, Vnode } from 'hslayout';
import { Popup } from './Popup';


export class EditLabel {
    protected editable = false;
    protected updateCB:(r:string) => void;

    protected blur(e:any) {
        this.editable = false;
        this.update(e.target.value);
    }

    protected click = () => this.editable = true; 

    protected keyup = (key:any) => key.which === 13? key.target.blur() : '';

    update(newValue:string) { this.updateCB(newValue); }

    onupdate(node:Vnode) {
        if (this.editable && document.activeElement!==node.dom) {
            node.dom.value = node.attrs.content || '';
            node.dom.focus();
            node.dom.select();
        }
    }

    view(node:Vnode) {
        this.updateCB = node.attrs.update;
        const css = node.attrs.css || '';
        const content = ''+node.attrs.content;
        return this.editable? m(`input.hsedit_label${css}`, { onblur:this.blur.bind(this), onkeyup:this.keyup.bind(this) }, 'yeah')
            : ( content &&  content.length)? 
            m(`span.hsedit_label${css}`, Popup.arm(node.attrs.popup, { onclick:()=>this.click() }), m.trust( content))
            : m(`span.hsedit_label.default${css}`, Popup.arm(node.attrs.popup, { onclick:()=>this.click() }), ''+node.attrs.placeholder || 'click to enter');
    }
}


/** an extension of `EditLabel` that pareses entries as dates. */
export class EditDate extends EditLabel {
    protected def = new Date().toDateString().slice(4);
    update(newValue:string) {
        if (newValue) {
            const date = new Date(newValue);
            const result = isNaN(date.getTime())? 'invalid date' : date.toDateString().slice(4);
            this.updateCB(result);
        } else {
            this.updateCB(undefined);
        }
    }
    public view(node:Vnode) {
        if (this.editable && (!node.attrs.content || node.attrs.content === '')) {
            node.attrs.content = new Date().toDateString();
            node.attrs.content = node.attrs.content.slice(4);
        }
        return super.view(node);
    }
}