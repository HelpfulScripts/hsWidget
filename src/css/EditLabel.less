
.hsedit_label {
    padding-top: 0px;
    padding-left: 7px;
    line-height: 1.5em;
    border: 0;  //3px solid rgba(0,0,0,0);
    border-top-width: 2px;
    input& {
        border:0;
        font-family: inherit;
        font-size: inherit;
        width:100%;
    }
    &.default {
        color: #aaa;
        font-style: italic;
    }
}