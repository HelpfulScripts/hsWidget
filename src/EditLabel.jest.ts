import { m, Vnode }     from 'hslayout';
import { EditLabel }    from './EditLabel';
import { Log }          from 'hsnode';  const log = new Log('EditLabel.jest');
import $                from 'jquery';

// window = Object.assign(require('mithril/test-utils/domMock.js')(), require('mithril/test-utils/pushStateMock')());
// const mq = require('mithril-query');

let label = '';

// describe('EditLabel', () => { 
//     const out = mq(m(EditLabel, {
//         css: '#myLabel',
//         placeholder: 'favorite hero',
//         update: (item:string) => label = item,
//         content: label
//     }));
//     it ('should have DOM structure', () => {
//         out.should.have('span.hsedit_label');
//         out.click('#myLabel');
//         const first = out.first('#myLabel');
//         out.click('#myLabel');
//         out.should.have('input');
//     });
// });

// describe('EditDate', () => { 
//     console.log(`EditDate...`);
//     const EditDate = require('./').EditDate;
//     let out = mq(m(EditDate, {
//         placeholder: 'favorite hero',
//         update: (item:string) => label = item,
//         content: label
//     }));
//     it ('should have DOM structure', () => {
//         out.should.have('span.hsedit_label');
//     });
// });

const root = window.document.createElement("div");

describe('EditLabel', () => {
    beforeAll(()=>{
        m.mount(root, { view: () => m(EditLabel, {
            css: '#myLabel',
            placeholder: 'favorite hero',
            update: (item:string) => label = item,
            content: label
        })});
    });

    it('should match snapshot', () => {
        expect(root).toMatchSnapshot();
    });
});