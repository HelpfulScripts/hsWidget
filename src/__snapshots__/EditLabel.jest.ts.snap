// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditLabel should match snapshot 1`] = `
<div>
  <span
    class="hsedit_label default"
    id="myLabel"
  >
    favorite hero
  </span>
</div>
`;
